File: src/claude_bridge/on_complete.py
#!/usr/bin/env python3
"""Stop hook handler — called by Claude Code when an agent task completes.

This script is referenced in agent .md frontmatter:
  hooks:
    Stop:
      - hooks:
          - type: command
            command: "python3 ~/.claude-bridge/on-complete.py --session-id <id>"
"""

import argparse
import json
import os
import sys
from datetime import datetime

# Add parent to path so we can import from the package
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from claude_bridge.db import BridgeDB


def parse_result_file(result_file: str) -> dict | None:
    """Parse the JSON result file from claude --output-format json."""
    try:
        expanded = os.path.expanduser(result_file)
        if not os.path.isfile(expanded):
            return None
        with open(expanded) as f:
            content = f.read().strip()
            if not content:
                return None
            return json.loads(content)
    except (json.JSONDecodeError, IOError):
        return None


def main(db: BridgeDB | None = None):
    parser = argparse.ArgumentParser(description="Claude Bridge stop hook handler")
    parser.add_argument("--session-id", required=True, help="Session ID of the completed task")
    args = parser.parse_args()

    own_db = db is None
    if own_db:
        db = BridgeDB()

    try:
        # Find the running task for this session
        task = db.get_running_task(args.session_id)
        if not task:
            return

        task_id = task["id"]
        result_file = task["result_file"]

        # Parse result
        status = "done"
        summary = ""
        cost = 0.0
        duration = 0
        turns = 0
        exit_code = 0
        error = None

        result = parse_result_file(result_file) if result_file else None

        if result:
            if result.get("is_error"):
                status = "failed"
                error = str(result.get("result", "Unknown error"))[:500]
            else:
                summary = str(result.get("result", ""))[:500]
            cost = result.get("cost_usd", 0) or 0
            duration = result.get("duration_ms", 0) or 0
            turns = result.get("num_turns", 0) or 0
        else:
            # No result file — check stderr
            if result_file:
                stderr_file = result_file.replace("-result.json", "-stderr.log")
                stderr_path = os.path.expanduser(stderr_file)
                if os.path.isfile(stderr_path):
                    with open(stderr_path) as f:
                        stderr_content = f.read().strip()
                    if stderr_content:
                        status = "failed"
                        error = stderr_content[:500]
                        exit_code = -1

        # Update task
        db.update_task(
            task_id,
            status=status,
            result_summary=summary if summary else None,
            cost_usd=cost,
            duration_ms=duration,
            num_turns=turns,
            exit_code=exit_code,
            error_message=error,
            completed_at=datetime.now().isoformat(),
        )

        # Update agent state
        db.update_agent_state(args.session_id, "idle")
        db.increment_agent_tasks(args.session_id)

        # Print report (Bridge Bot picks this up)
        if duration:
            mins = duration // 60000
            secs = (duration % 60000) // 1000
            duration_str = f"{mins}m {secs}s"
        else:
            duration_str = "unknown"

        if status == "done":
            print(f"✓ Task #{task_id} ({args.session_id}) — done in {duration_str}")
            if summary:
                print(f"  {summary[:200]}")
            print(f"  Cost: ${cost:.3f} | Turns: {turns}")
        else:
            print(f"✗ Task #{task_id} ({args.session_id}) — failed after {duration_str}")
            if error:
                print(f"  Error: {error[:200]}")

    finally:
        if own_db:
            db.close()


if __name__ == "__main__":
    main()

File: src/execution/on-complete.ts
/**
 * Completion Handler — stop hook callback when Claude Code finishes.
 *
 * Reads result files, updates task in DB, sends notification, dequeues next task.
 * Matches Python on_complete.py behavior.
 */

import { existsSync, readFileSync } from "fs";
import type { ICompletionHandler, CompletionResult, IDispatcher } from "./interfaces.js";
import type { IDatabase } from "../data/interfaces.js";
import { startTask } from "./dispatcher.js";

/**
 * Callback fired when a completed task belongs to an active loop. Decouples
 * CompletionHandler from LoopOrchestrator so the execution layer doesn't
 * import from orchestration. Callers wire this explicitly.
 */
export type LoopCompletionCallback = (
  loopId: string,
  taskId: number,
  resultSummary: string,
  costUsd: number,
) => Promise<void>;

export class CompletionHandler implements ICompletionHandler {
  constructor(
    private homeDir: string,
    private db: IDatabase,
    private dispatcher?: IDispatcher,
    private onLoopTaskComplete?: LoopCompletionCallback,
  ) {}

  async parseResultFile(resultFile: string): Promise<CompletionResult | null> {
    if (!existsSync(resultFile)) return null;

    try {
      const raw = readFileSync(resultFile, "utf-8");
      const data = JSON.parse(raw) as Record<string, unknown>;

      // claude's --output-format json uses `total_cost_usd`; older/legacy
      // payloads may use `cost_usd`. Accept either.
      const cost = (data["total_cost_usd"] as number | undefined)
        ?? (data["cost_usd"] as number | undefined)
        ?? null;
      return {
        exitCode: data["is_error"] ? 1 : 0,
        summary: (data["result"] as string | undefined) ?? null,
        costUsd: cost,
        durationMs: (data["duration_ms"] as number | undefined) ?? null,
        numTurns: (data["num_turns"] as number | undefined) ?? null,
      };
    } catch {
      return null;
    }
  }

  async handleCompletion(
    sessionId: string,
    taskId: number,
    result: CompletionResult,
  ): Promise<void> {
    const task = this.db.getTask(taskId);
    if (!task) return;

    const isSuccess = result.exitCode === 0;
    const now = new Date().toISOString();

    // Update task
    this.db.updateTask(taskId, {
      status: isSuccess ? "done" : "failed",
      result_summary: isSuccess ? result.summary : undefined,
      error_message: isSuccess ? undefined : (result.summary ?? "Unknown error"),
      cost_usd: result.costUsd ?? undefined,
      duration_ms: result.durationMs ?? undefined,
      num_turns: result.numTurns ?? undefined,
      exit_code: result.exitCode,
      completed_at: now,
    });

    // Update agent state to idle
    this.db.updateAgentState(sessionId, "idle");

    // Create notification if channel_chat_id is set
    if (task.channel_chat_id) {
      const status = isSuccess ? "done" : "failed";
      const summary = result.summary ?? "No summary";
      const costStr = result.costUsd !== null ? ` ($${result.costUsd.toFixed(2)})` : "";
      const message = `Task #${taskId} ${status}${costStr}: ${summary}`;
      this.db.createNotification(taskId, task.channel, task.channel_chat_id, message);
    }

    // If this task belonged to an active goal loop, let the orchestrator
    // decide the next step. The orchestrator may dispatch the next iteration
    // (which spawns a new task and flips the agent back to `running`), finish
    // the loop, or pause for manual approval.
    const agent = this.db.getAgentBySession(sessionId);
    let loopHandled = false;
    if (agent && this.onLoopTaskComplete) {
      const loop = this.db.getActiveLoopForAgent(agent.name);
      if (loop && loop.current_task_id === String(taskId)) {
        try {
          await this.onLoopTaskComplete(
            loop.loop_id,
            taskId,
            result.summary ?? "",
            result.costUsd ?? 0,
          );
          loopHandled = true;
        } catch (err) {
          process.stderr.write(`[on-complete] loop callback failed: ${err}\n`);
        }
      }
    }

    // Only dequeue if no loop took the agent. When the loop dispatches the
    // next iteration it claims the agent itself; dequeuing on top would
    // double-dispatch.
    if (!loopHandled) {
      const next = this.db.dequeueNextTask(sessionId);
      if (next && this.dispatcher && agent) {
        try {
          await startTask(this.db, this.dispatcher, next, agent);
        } catch {
          /* startTask already recorded the failure */
        }
      }
    }
  }

  async main(argv?: string[]): Promise<void> {
    // CLI entry point for stop hook
    // Expected args: session_id task_id [result_file]
    const args = argv ?? process.argv.slice(2);
    if (args.length < 2) {
      process.stderr.write("Usage: on-complete <session_id> <task_id> [result_file]\n");
      process.exit(1);
    }

    const sessionId = args[0]!;
    const taskId = parseInt(args[1]!, 10);
    const resultFile = args[2];

    let result: CompletionResult = {
      exitCode: 0,
      summary: null,
      costUsd: null,
      durationMs: null,
      numTurns: null,
    };

    if (resultFile) {
      const parsed = await this.parseResultFile(resultFile);
      if (parsed) result = parsed;
    }

    await this.handleCompletion(sessionId, taskId, result);
  }
}
